Add tiling documentation (#2204)
open-edge-platform/training_extensions: docs/source/guide/explanation/additional_features/tiling.rst

Improve Small Object Detection with Image Tiling
================================================

The OpenVINO Training Extension introduces the concept of image tiling to improve the accuracy of detection algorithms and instance segmentation algorithms, particularly for small and densely packed objects in high-resolution images.

Image tiling involves dividing the original full-resolution image into multiple smaller tiles or patches. This division allows objects within the tiles to appear larger in relation to the tile size, effectively addressing the challenge of objects becoming nearly invisible in deeper layers of feature maps due to downsampling operations. For datasets where objects can be as small as 20 by 20 pixels in a 4K image, image tiling proves especially beneficial.

> However, it's important to consider the trade-off associated with image tiling. By dividing a single image sample into several tiles, the number of samples for training, evaluation, and testing increases. This trade-off impacts the speed of execution, as processing more images requires additional computational resources. To strike a balance between patch size and computational efficiency, the OpenVINO Training Extension incorporates tile dataset samples and adaptive tiling parameter optimization. These features enable the proper tuning of tile size and other tiling-related parameters to ensure efficient execution without compromising accuracy.

Reviewer: "tile dataset samples and adaptive tiling parameter optimization" - maybe add more insights about that algorithms and explain them deeper?

Author: I added a new section: Tile Size and Tile Overlap Optimization

> However, it's important to consider the trade-off associated with image tiling. By dividing a single image sample into several tiles, the number of samples for training, evaluation, and testing increases. This trade-off impacts the speed of execution, as processing more images requires additional computational resources. To strike a balance between patch size and computational efficiency, the OpenVINO Training Extension incorporates tile dataset samples and adaptive tiling parameter optimization. These features enable the proper tuning of tile size and other tiling-related parameters to ensure efficient execution without compromising accuracy.

Reviewer 2: OpenVINO Training Extension -> OpenVINO Training ExtensionS
Can you please fix it in other usages as well please?

Reviewer 3: > OpenVINO Training Extension -> OpenVINO Training ExtensionS Can you please fix it in other usages as well please?

Should we use capital `S` ..?

Reviewer 2: @sungmanc No, I just made it capital to emphasize the change

By leveraging image tiling, the OpenVINO Training Extension empowers detection and instance segmentation algorithms to effectively detect and localize small and crowded objects in large-resolution images, ultimately leading to improved overall performance and accuracy.

Tiling Strategies 
**********************************
Below we provided an example of tiling used on one of the image from `DOTA <https://captain-whu.github.io/DOTA/dataset.html>`_.

.. image:: ../../../../utils/images/dota_tiling_example.jpg
  :width: 800
  :alt: this image uploaded from this `source <https://captain-whu.github.io/DOTA/dataset.html>`_

In this example, the full image is cropped into 9 tiles. During training, only the tiles with annotations (bounding boxes or masks) are used for training.

During evaluation in training, only the tiles with annotations are used for evaluation, and evaluation is performed at the tile level.

> During testing, each tile is processed and predicted separately. The tiles are then stitched back together to form the full image, and the tile predictions are merged to form the full image prediction.

Reviewer 2: Should we mention here that we also infer the model on full image to detect well both large objects and small ones on tiles?

Reviewer 2: We can also mention that tiles cover images with overlap to ensure the detection of objects around borders.

Author: I mentioned it in `Tile Size and Tile Overlap Optimization`. Let me know if it's not clear enough.

Reviewer 2: Thank you, it's all clear now

The tiling strategy is implemented in the OpenVINO Training Extension through the following steps:

.. code-block:: 

    * Training: Create an ImageTilingDataset with annotated tiles -> Train with annotated tile images -> Evaluate on annotated tiles
    * Testing: Create an ImageTilingDataset including all tiles -> Test with all tile images -> Stitching -> Merge tile-level predictions -> Full Image Prediction

.. note::

    While running `ote eval` on models trained with tiling enabled, the evaluation will be performed on all tiles, this process including mergeing all the tile-level prediction. 
    The below context will be provided during evaluation:

    .. code-block:: 

        [>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 650/650, 17.2 task/s, elapsed: 38s, ETA:     0s
        ==== merge: 7.326097726821899 sec ====

Enable Tiling via OTX Training CLI 
**********************************

Currently tiling is supported for both detection and instance segmentation models.

> .. code-block:: 

Reviewer: I don't think we need to add output of the templates for OD and IS. We can mention, that tiling is available for these tasks and put the links to explanation sections of OD and IS

Reviewer 3: Agree with this opinion.

    |          TASK         |                               ID                               |           NAME           |                                           BASE PATH                                           |
    +-----------------------+----------------------------------------------------------------+--------------------------+-----------------------------------------------------------------------------------------------+
    | INSTANCE_SEGMENTATION |    Custom_Counting_Instance_Segmentation_MaskRCNN_ResNet50     |    MaskRCNN-ResNet50     |     otx/algorithms/detection/configs/instance_segmentation/resnet50_maskrcnn/template.yaml    |
    | INSTANCE_SEGMENTATION | Custom_Counting_Instance_Segmentation_MaskRCNN_EfficientNetB2B | MaskRCNN-EfficientNetB2B | otx/algorithms/detection/configs/instance_segmentation/efficientnetb2b_maskrcnn/template.yaml |
    +-----------------------+----------------------------------------------------------------+--------------------------+-----------------------------------------------------------------------------------------------+

> To enable tiling in OTX training, set the tiling_parameters.enable_tiling parameter to 1. Here's an example of enabling tiling for the SSD model template:

Reviewer 2: ```suggestion
To enable tiling in OTX training, set the ``tiling_parameters.enable_tiling`` parameter to 1. Here's an example of enabling tiling for the SSD model template:
```

.. code-block::

>     otx train Custom_Object_Detection_Gen3_SSD --train-data-roots tests/assets/small_objects --val-data-roots tests/assets/small_objects params --tiling_parameters.enable_tiling 1

Reviewer: I suggest combining this example and "Customizing Tiling Parameters" in one section. Also, I would rather remove the "Run Tiling on OpenVINO Exported Model" section since no difference between ordinary models and we already have tutorials on that.

Reviewer 2: I personally don't have problems with "Run Tiling on OpenVINO Exported Model" section, it's a duplication, yes, but it's convenient then all info related to the Tiling inference is in the same place

Reviewer 2: Can you please add another option, where we build a workplace and then manually update the ``template.yaml`` to use tiling and update tiling parameters?

`otx build Custom_Object_Detection_Gen3_SSD --train-data-roots tests/assets/small_objects --val-data-roots tests/assets/small_objects params 

cd workspace-DETECTION

*update template*
train 
`

Reviewer 2: I would add here an option with otx build as well to make calling other train function easier and concise, like that:

```
Here's an example of enabling tiling for the SSD model template using workspace:

otx build Custom_Object_Detection_Gen3_SSD --train-data-roots tests/assets/small_objects --val-data-roots tests/assets/small_objects params

cd workspace-DETECTION

otx train params --tiling_parameters.enable_tiling 1 ...

Also we can update ``tiling_parameters`` adding them in ``template.yaml`` like that:

hyper_parameters:
...
    tiling_parameters:
      enable_tiling:
        default_value: true

```

Author: @GalyaZalesskaya  I added a new section: `Enable Tiling via OTX Build`

    params --tiling_parameters.enable_tiling 1          \  # enable tiling
           --tiling_parameters.enable_adaptive_params 0 \  # disable automatic tiling parameter optimization
           --tiling_parameters.tile_size 512 \             # tile size configured to 512x512
           --tiling_parameters.tile_overlap 0.1 \          # 10% overlap between tiles

By specifying these parameters, you can disable automatic tiling parameter optimization, configure the tile size to 512x512, and set a 10% overlap between tiles.

Run Tiling on OpenVINO Exported Model
**************************************

After training a model with tiling enabled, you can export the model to OpenVINO IR format using the following command:

.. code-block:: 

    otx export Custom_Object_Detection_Gen3_SSD --load-weights <path_to_trained_model>/weights.pth --output <path_to_exported_model>

After exporting the model, you can run inference on the exported model using the following command:

.. code-block:: 

    ote eval Custom_Object_Detection_Gen3_SSD --test-data-roots tests/assets/small_objects --load-weights <path_to_exported_model>/openvino.xml

> .. warning::
>     Tiling trades off speed for accuracy as it increases the number of images to be processed. Therefore, it is normal to see a longer training and inference time when tiling is enabled.

Reviewer 2: ```suggestion
.. warning::
    
    Tiling trades off speed for accuracy as it increases the number of images to be processed. Therefore, it is normal to see a longer training and inference time when tiling is enabled.
```

>     Tiling trades off speed for accuracy as it increases the number of images to be processed. Therefore, it is normal to see a longer training and inference time when tiling is enabled.

Reviewer 2: Also we can make another note that sometimes batchsize decrease is required to train on  machines that are not really powerful

.. image:: ../../../../utils/images/dota_tiling_example.jpg
  :width: 800
  :alt: this image uploaded from this `source <https://captain-whu.github.io/DOTA/dataset.html>`_

In this example, the full image is cropped into 9 tiles. During training, only the tiles with annotations (bounding boxes or masks) are used for training.

During evaluation in training, only the tiles with annotations are used for evaluation, and evaluation is performed at the tile level.

During testing, each tile is processed and predicted separately. The tiles are then stitched back together to form the full image, and the tile predictions are merged to form the full image prediction.

The tiling strategy is implemented in the OpenVINO Training Extensions through the following steps:

.. code-block:: 

    * Training: Create an ImageTilingDataset with annotated tiles -> Train with annotated tile images -> Evaluate on annotated tiles
    * Testing: Create an ImageTilingDataset including all tiles -> Test with all tile images -> Stitching -> Merge tile-level predictions -> Full Image Prediction

.. note::

>     While running `ote eval` on models trained with tiling enabled, the evaluation will be performed on all tiles, this process including mergeing all the tile-level prediction. 

Reviewer 2: ```suggestion
    While running `ote eval` on models trained with tiling enabled, the evaluation will be performed on all tiles, this process includes merging all the tile-level predictions. 
```

Adaptive tiling parameter optimization works by finding the average object size in the training dataset and using that to determine the tile size. Currently, the average object size to tile size ratio is set to 3%. For example, if the average object size is 100x100 pixels, the tile size will be around 577x577 pixels.

This computation is performed by dividing the average object size by the desired object size ratio (default: 3%) and then taking the square root. This ensures that the objects are large enough to be detected by the model. The object size to tile size ratio can also be configured with ``tiling_parameters.object_tile_ratio`` parameter. 

Here's an example of setting the object size ratio to 5%:

.. code-block:: 
    
    otx train Custom_Object_Detection_Gen3_SSD --train-data-roots tests/assets/small_objects --val-data-roots tests/assets/small_objects
    params --tiling_parameters.enable_tiling 1          \  # enable tiling
           --tiling_parameters.enable_adaptive_params 1 \  # enable automatic tiling parameter optimization

>            --tiling_parameters.object_tile_ratio 0.05   \  # set the object size ratio to 5%

Reviewer 2: If we'd use workspace creation how we do in most of the tutorials now, we'd make this line shorter:

    otx train \
    params --tiling_parameters.enable_tiling 1          \  # enable tiling
           --tiling_parameters.enable_adaptive_params 1 \  # enable automatic tiling parameter optimization
           --tiling_parameters.object_tile_ratio 0.05   \  # set the object size ratio to 5%

Reviewer 2: But if we'd keep such example, it's better to reformat it a little:

```
 otx train Custom_Object_Detection_Gen3_SSD 
    --train-data-roots tests/assets/small_objects \
    --val-data-roots tests/assets/small_objects \
    params --tiling_parameters.enable_tiling 1          \  # enable tiling
           --tiling_parameters.enable_adaptive_params 1 \  # enable automatic tiling parameter optimization
           --tiling_parameters.object_tile_ratio 0.05   \  # set the object size ratio to 5%
```
